YodaStyleFixer - Add always_move_variable option

julienfalque opened the pull request:
Closes #3153. Should be merged after #3674.

SpacePossum commented on src/Fixer/Phpdoc/PhpdocAlignFixer.php:
```diff
@@ -228,7 +228,7 @@ private function fixDocBlock($content)
             foreach ($items as $j => $item) {
                 if (null === $item['tag']) {
-                    if ($item['desc'][0] === '@') {
+                    if ('@' === $item['desc'][0]) {
```
I think this change is good and indeed shows the bug you are fixing in the other PR

julienfalque commented on src/Fixer/ControlStructure/YodaStyleFixer.php:
```diff
@@ -114,6 +119,14 @@ protected function createConfigurationDefinition()
                 ->getOption(),
+            (new FixerOptionBuilder('always_move_variable', 'Whether variables should always be on non assignable side when applying Yoda style.'))
+                ->setAllowedTypes(['bool'])
+                ->setDefault(false)
```
By the way I would enable these options in our `.php_cs.dist` file and set their default value to `true` on 3.0 (or even drop the options). WDYT?

SpacePossum replied:
I would like these new defaults, however I would not drop the options as it would stop SF from updating to 3.0 I think, but that is just me guessing.

julienfalque commented:
Rebased.

SpacePossum commented:
Please add a code sample showing how the new options work.

SpacePossum commented on src/Fixer/ControlStructure/YodaStyleFixer.php:
```diff
@@ -481,17 +506,35 @@ private function isOfLowerPrecedence(Token $token)
+
+            for ($index = $start; $index <= $end; ++$index) {
+                if (
+                    $tokens[$index]->isGivenKind([T_POW, T_IS_SMALLER_OR_EQUAL, T_IS_GREATER_OR_EQUAL, T_SL, T_SR, T_INSTANCEOF, T_BOOL_CAST, T_OBJECT_CAST, T_ARRAY_CAST, T_STRING_CAST, T_DOUBLE_CAST, T_INT_CAST])
```
can we use `TokensAnalyzer::isBinaryOperator` here?

SpacePossum commented on src/Fixer/ControlStructure/YodaStyleFixer.php:
```diff
@@ -598,7 +646,26 @@ private function isVariable(Tokens $tokens, $index, $end)
+
+        // -1 or -1.2
+        if ($index === $end - 1) {
+            return $token->equals('-') && $tokens[$end]->isGivenKind([T_LNUMBER, T_DNUMBER]);
```
I think `$tokens[$end]` might a comment/PHPDoc here; `$a/**/=/**/-/**/1/**/;`

julienfalque commented:
> Please add a code sample showing how the new options work.

Done.

I also removed the `always_move_constant` option I introduced, I noticed it actually has no effect: constants are moved even without the option. Not sure why I introduced it. Maybe something fixed in upstream branches.

SpacePossum commented on src/Fixer/ControlStructure/YodaStyleFixer.php:
```diff
@@ -481,17 +506,35 @@ private function isOfLowerPrecedence(Token $token)
+        }
+
+        if ($strict) {
+            if ($tokens[$start]->equals('(')) {
```
I was thinking about cases like `((( ($foo) === (1))))` or any construct that is supported within `$strict` with the meaning less extra `()` added. Maybe I miss understand, but the new options will not support fixing these cases by design?

SpacePossum replied:
I think for `if (  (($foo))) === ((1)) )` it is called twice for `(($foo))` and `((1))` and not for `($foo)` and `(1)`, but I've to re-check TBH :)

keradus commented:
Thank you @julienfalque.
